File: tiledb/sys2/cpp/layout.cc
#include <string>

#include <tiledb/tiledb.h>

#include "exception.h"
#include "tiledb-sys2/src/layout.rs.h"

namespace tiledb::rs {

tiledb_layout_t to_cpp_tile_order(TileOrder order) {
  switch (order) {
    case TileOrder::RowMajor:
      return TILEDB_ROW_MAJOR;
    case TileOrder::ColumnMajor:
      return TILEDB_COL_MAJOR;
    default:
      throw TileDBError("Invalid TileOrder for conversion.");
  }
}

TileOrder to_rs_tile_order(tiledb_layout_t order) {
  switch (order) {
    case TILEDB_ROW_MAJOR:
      return TileOrder::RowMajor;
    case TILEDB_COL_MAJOR:
      return TileOrder::ColumnMajor;
    default:
      throw TileDBError("Invalid tiledb_layout_t for TileOrder conversion.");
  }
}

tiledb_layout_t to_cpp_cell_order(CellOrder order) {
  switch (order) {
    case CellOrder::Unordered:
      return TILEDB_UNORDERED;
    case CellOrder::RowMajor:
      return TILEDB_ROW_MAJOR;
    case CellOrder::ColumnMajor:
      return TILEDB_COL_MAJOR;
    case CellOrder::Global:
      return TILEDB_GLOBAL_ORDER;
    case CellOrder::Hilbert:
      return TILEDB_HILBERT;
    default:
      throw TileDBError("Invalid CellOrder for conversion.");
  }
}

CellOrder to_rs_cell_order(tiledb_layout_t order) {
  switch (order) {
    case TILEDB_UNORDERED:
      return CellOrder::Unordered;
    case TILEDB_ROW_MAJOR:
      return CellOrder::RowMajor;
    case TILEDB_COL_MAJOR:
      return CellOrder::ColumnMajor;
    case TILEDB_GLOBAL_ORDER:
      return CellOrder::Global;
    case TILEDB_HILBERT:
      return CellOrder::Hilbert;
    default:
      throw TileDBError("Invalid tiledb_layout_t for CellOrder conversion.");
  }
}

}  // namespace tiledb::rs

File: tiledb/sys2/src/layout.rs
use tiledb_common::array::{CellOrder, TileOrder};

use crate::error::TryFromFFIError;

#[cxx::bridge(namespace = "tiledb::rs")]
mod ffi {
    #[derive(Debug)]
    pub enum TileOrder {
        RowMajor,
        ColumnMajor,
    }

    #[derive(Debug)]
    pub enum CellOrder {
        Unordered,
        RowMajor,
        ColumnMajor,
        Global,
        Hilbert,
    }
}

pub use ffi::CellOrder as FFICellOrder;
pub use ffi::TileOrder as FFITileOrder;

impl From<TileOrder> for FFITileOrder {
    fn from(order: TileOrder) -> FFITileOrder {
        match order {
            TileOrder::RowMajor => FFITileOrder::RowMajor,
            TileOrder::ColumnMajor => FFITileOrder::ColumnMajor,
        }
    }
}

impl TryFrom<FFITileOrder> for TileOrder {
    type Error = TryFromFFIError;

    fn try_from(order: FFITileOrder) -> Result<Self, Self::Error> {
        let order = match order {
            FFITileOrder::RowMajor => TileOrder::RowMajor,
            FFITileOrder::ColumnMajor => TileOrder::ColumnMajor,
            _ => return Err(TryFromFFIError::from_tile_order(order)),
        };
        Ok(order)
    }
}

impl From<CellOrder> for FFICellOrder {
    fn from(order: CellOrder) -> FFICellOrder {
        match order {
            CellOrder::Unordered => FFICellOrder::Unordered,
            CellOrder::RowMajor => FFICellOrder::RowMajor,
            CellOrder::ColumnMajor => FFICellOrder::ColumnMajor,
            CellOrder::Global => FFICellOrder::Global,
            CellOrder::Hilbert => FFICellOrder::Hilbert,
        }
    }
}

impl TryFrom<FFICellOrder> for CellOrder {
    type Error = TryFromFFIError;

    fn try_from(order: FFICellOrder) -> Result<Self, Self::Error> {
        let order = match order {
            FFICellOrder::Unordered => CellOrder::Unordered,
            FFICellOrder::RowMajor => CellOrder::RowMajor,
            FFICellOrder::ColumnMajor => CellOrder::ColumnMajor,
            FFICellOrder::Global => CellOrder::Global,
            FFICellOrder::Hilbert => CellOrder::Hilbert,
            _ => return Err(TryFromFFIError::from_cell_order(order)),
        };
        Ok(order)
    }
}
